src/o365/spo/commands/page/page-clientsidewebpart-add.ts

```diff
@@ -0,0 +1,483 @@
+import auth from '../../SpoAuth';
+import config from '../../../../config';
+import * as request from 'request-promise-native';
+import commands from '../../commands';
+import { CommandOption, CommandValidate } from '../../../../Command';
+import SpoCommand from '../../SpoCommand';
+import Utils from '../../../../Utils';
+import GlobalOptions from '../../../../GlobalOptions';
+import { Auth } from '../../../../Auth';
+import {
+	ClientSidePage,
+	CanvasSection,
+	ClientSideWebpart,
+	ClientSidePageComponent,
+	CanvasColumn
+} from './clientsidepages';
+import { StandardWebPart, StandartWebPartUtils } from './StandardWebpartTypes';
+import { ContextInfo } from '../../spo';
+
+const vorpal: Vorpal = require('../../../../vorpal-init');
+
+interface CommandArgs {
+	options: Options;
+}
+
+interface Options extends GlobalOptions {
+	pageName: string;
+	webUrl: string;
+	standardWebPart: StandardWebPart;
+	webPartId: string;
+	webPartProperties: string;
+	section: number;
+	column: number;
+	order: number;
+}
+
+class SpoPageClientSideWebPartAddCommand extends SpoCommand {
+	public get name(): string {
+		return `${commands.PAGE_CLIENTSIDEWEBPART_ADD}`;
+	}
+
+	public get description(): string {
+		return 'Adds a client side WebPart to a modern page';
+	}
+
+	public commandAction(cmd: CommandInstance, args: CommandArgs, cb: (err?: any) => void): void {
+		const resource: string = Auth.getResourceFromUrl(args.options.webUrl);
+		let siteAccessToken: string = '';
+		let requestDigest: string = '';
+		let clientSidePage: ClientSidePage = null as any;
+
+		let pageName: string = args.options.pageName;
+		if (args.options.pageName.indexOf('.aspx') < 0) {
+			pageName += '.aspx';
+		}
+
+		if (this.debug) {
+			cmd.log(`Retrieving access token for ${resource}...`);
+		}
+
+		auth
+			.getAccessToken(resource, auth.service.refreshToken as string, cmd, this.debug)
+			.then((accessToken: string): request.RequestPromise => {
+				if (this.debug) {
+					cmd.log(`Retrieved access token ${accessToken}. Retrieving request digest...`);
+				}
+
+				siteAccessToken = accessToken;
+
+				if (this.verbose) {
+					cmd.log(`Retrieving request digest...`);
+				}
+
+				return this.getRequestDigestForSite(args.options.webUrl, siteAccessToken, cmd, this.debug);
+			})
+			.then((res: ContextInfo): Promise<ClientSidePage> => {
+				if (this.debug) {
+					cmd.log('Response:');
+					cmd.log(res);
+					cmd.log('');
+				}
+
+				// Keep the reference of request digest for subsequent requests
+				requestDigest = res.FormDigestValue;
+
+				if (this.verbose) {
+					cmd.log(`Retrieving Client Side Page ${pageName}...`);
+				}
+				// Get Client Side Page
+				return this._getClientSidePage(pageName, cmd, args, siteAccessToken);
+			})
+			.then((csPage) => {
+				if (this.debug) {
+					cmd.log(`Retrieved Client Side Page:`);
+					cmd.log(csPage);
```

Comment:
When running the command with the `--debug` option, the command fails at this line with the following error:

```
Retrieved Client Side Page:
Error: Converting circular structure to JSON
```

Since we're already including the retrieved page contents in the `_getClientSidePage` method, it's not necessary to do it again here, so this could be removed.